dwa_local_planner/src/dwa_planner_ros.cpp

```diff
@@ -209,7 +213,28 @@ namespace dwa_local_planner {
     delete dsrv_;
   }
 
+  void DWAPlannerROS::resetBestEffort() {
+    latched_inner_goal_ = false;
+    prev_vel_dir_ = 0;
+    oscillating_ = false;
+  }
+
+  bool DWAPlannerROS::finishedBestEffort() {
+    if (transformed_plan_.empty()) {
+      return false;
+    }
 
+    // latch inner tolerance
+    const double goal_x = transformed_plan_.back().pose.position.x;
+    const double goal_y = transformed_plan_.back().pose.position.y;
+    const double inner_xy_goal_tolerance = planner_util_.getCurrentLimits().inner_xy_goal_tolerance;
+    const double goal_dist = base_local_planner::getGoalPositionDistance(current_pose_, goal_x, goal_y);
+    latched_inner_goal_ = latched_inner_goal_ || goal_dist <= inner_xy_goal_tolerance;
+
+    const bool bypassed_goal = base_local_planner::isGoalBypassed(transformed_plan_, current_pose_);
```

Comment:
Q: I am wondering a bit whether this is a actually a useful check, because the global plan updates all the time, making the robot being in front of the goal again?

S: from looking at the current `bypassed_goal`, `vel_dir` and `cmd_vel` logic, I think that could be simplified into a single robust metric: Checking whether the distance to the goal is decreasing?

Reply:
> because the global plan updates all the time, making the robot being in front of the goal again?

Correct, but this doesn't happen instantly, so the first time we overshoot it should still be detected. This is easy to test by teleporting the robot around -- it will detect the overshoot for a few cycles and then go back to not overshooting once the global plan updates.

>  Checking whether the distance to the goal is decreasing?

Me, @ayushgaud and @corot discussed this approach and decided it would be better to not do it.
There are cases where the distance to the goal is still decreasing but we've already bypassed it.
Old version of path follower used this approach, and was later changed to the current one (check vel direction + bypass) due to this problem.

Reply:
@ChristofDubs implemented the thing we discussed yesterday about using the point when we first entered the outer tolerance rather than the 3rd last pose of the path. But we still need access to the goal pose, so we still need to store the transformed plan (well, technically I guess we could just store the transformed goal, but meh same thing).